File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        
        self.index = 0
        
        def helper(left_boundary, right_boundary):   
            
            if self.index >= len(preorder):        
                return None
            
            val = preorder[self.index]              
            
            # check for boundary condition
            if val < left_boundary or val > right_boundary:
                return None                         
            
            root = TreeNode(val)                    
            self.index += 1                         
            root.left = helper(left_boundary, val)  
            root.right = helper(val, right_boundary) 
            
            return root
        
        return helper(float("-inf"), float("inf"))
        
            
            

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] preorder;
    public int index;
    
    public TreeNode helper(int lower, int upper) {
        if (index == preorder.length)
            return null;
        
        int value = preorder[index];
        
        if (value < lower || value > upper)
            return null;
        
        index++;
        TreeNode root = new TreeNode(value);
        root.left = helper(lower, value);
        root.right = helper(value, upper);
        
        return root;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        this.preorder = preorder;
        index = 0;
        return helper(Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}